aws/resource_aws_storagegateway_fsx_associate_file_system_test.go

```diff
@@ -0,0 +1,374 @@
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAwsStorageGatewayFsxAssociateFileSystemConfig_Cache(rName, username, 400),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsStorageGatewayFsxAssociateFileSystemExists(resourceName, &fsxFileSystemAssociation),
+					resource.TestCheckResourceAttr(resourceName, "cache_attributes.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "cache_attributes.0.cache_stale_timeout_in_seconds", "400"),
+				),
+			},
+			{
+				ResourceName:            resourceName,
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"username", "password"},
+			},
+			{
+				Config: testAccAwsStorageGatewayFsxAssociateFileSystemConfig_Cache(rName, username, 0),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsStorageGatewayFsxAssociateFileSystemExists(resourceName, &fsxFileSystemAssociation),
+					resource.TestCheckResourceAttr(resourceName, "cache_attributes.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "cache_attributes.0.cache_stale_timeout_in_seconds", "0"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSStorageGatewayFsxAssociateFileSystem_auditDestination(t *testing.T) {
+	var fsxFileSystemAssociation storagegateway.FileSystemAssociationInfo
+	rName := acctest.RandomWithPrefix("tf-acc-test")
+	resourceName := "aws_storagegateway_fsx_associate_file_system.test"
+	username := "Admin"
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t); testAccPartitionHasServicePreCheck(storagegateway.EndpointsID, t) },
+		ErrorCheck:   testAccErrorCheck(t, storagegateway.EndpointsID),
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAwsStorageGatewayFsxAssociateFileSystemDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAwsStorageGatewayFsxAssociateFileSystemConfig_Audit(rName, username, "aws_cloudwatch_log_group.test.arn"),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsStorageGatewayFsxAssociateFileSystemExists(resourceName, &fsxFileSystemAssociation),
+					resource.TestCheckResourceAttrPair(resourceName, "audit_destination_arn", "aws_cloudwatch_log_group.test", "arn"),
+				),
+			}, {
+				ResourceName:            resourceName,
+				ImportState:             true,
+				ImportStateVerify:       true,
+				ImportStateVerifyIgnore: []string{"username", "password"},
+			},
+			{
+				Config: testAccAwsStorageGatewayFsxAssociateFileSystemConfig_Audit(rName, username, "aws_cloudwatch_log_group.test2.arn"),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsStorageGatewayFsxAssociateFileSystemExists(resourceName, &fsxFileSystemAssociation),
+					resource.TestCheckResourceAttrPair(resourceName, "audit_destination_arn", "aws_cloudwatch_log_group.test2", "arn"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSStorageGatewayFsxAssociateFileSystem_disappears(t *testing.T) {
```

Comment:
can we add a check for SGW/FSX disappears as well and how the association handles it? (if there invalid then no need but sometimes when its valid other error messages occur and we want to handle those)

Reply:
How would you propose that we issue a delete command to a FSx file system, and/or a Gateway from the AccTest? We are not sure how we would go about that.

Reply:
see example in APIGW AccTests where in the policy resource the APIGW it self is deleted:

https://github.com/hashicorp/terraform-provider-aws/blob/ac06ced75cba0daf09fef2938752ad13cc6fff6e/aws/resource_aws_api_gateway_rest_api_policy_test.go#L88

Reply:
from the AWS console:
if you delete the sgw, the association is destroyed
if you delete the fsx file system, the association reports Status: ERROR

for sgw deletes, obviously we want the provider to gracefully remove the association resource from state
for fsx file system deletes, what action do we want the provider to take? should we include Status in the state file and merely report back the ERROR?

Reply:
great instincts on these test scenarios, @DrFaust92! ok so i have tests for both now but that uncovered a bug in the API. deleting the storage gateway works as expected, no issue.

deleting the fsx file system, however, makes the association meta object (this resource) undeletable and a dangling resource. 
running `apply` once creates this error:
`│ Error: error waiting for Storage Gateway FSx File System Association ("arn:aws:storagegateway:us-west-2:<>:fs-association/fsa-065AFF7FD1ECC1A0A") to be deleted: timeout while waiting for resource to be gone (last state: 'DELETING', timeout: 3m0s)`
and then the resource is removed from the state file

that seems like desirable behavior to me? halt the process, report the error and remove the resource from state, do you agree?

Reply:
this is completed if you agree with the above, @DrFaust92